Add HttpServerErrorHandler interface to provide visibility and customization of protocol errors

harrishancock opened the pull request:
~~This avoids needing to add another option flag to control whether or not to log protocol errors. It's a bit more flexible, too -- we could use these hooks to log 400 responses, but ignore other errors or record them only as metrics.~~

Updated: separated the hook out into an interface and provided a `Response&` for response customization.

jclee commented on c++/src/kj/compat/http.h:
```diff
@@ -647,6 +647,18 @@ class HttpService {
+  // HttpService. This can happen when HttpServer encounters a protocol error, or if
+  // `HttpService::request()` throws an exception before writing a response. The parameters contain
+  // the information that HttpServer will send back in the HTTP response, and are provided for
+  // visibility only only, e.g. to allow applications to log or count such occurrences.
```
"only only"

harrishancock replied:
Rewritten in new version.

harrishancock commented:
Take two.

The `closeAfterSend` boolean is a hack to send a Connection: close header, the absence of which breaks this test: https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/compat/http-test.c%2B%2B#L3031

The "HttpService threw exception after generating a partial response ..." message changed a bit in a way I'm not super happy about. The logic at the place that generates that message is now inside of the default implementation of `HttpServerErrorHandler::requestError()` (a consequence of passing a `Maybe<Response&>` to indicate partial responses), meaning it no longer has direct access to the original exception. Instead, it wraps the HttpServer's "suggested" error message, which itself wraps the exception's message.

jclee commented on c++/src/kj/compat/http.c++:
```diff
@@ -5063,20 +5056,23 @@ private:
+                       kj::PromiseFulfillerPair<void> paf)
     : timer(timer), requestHeaderTable(requestHeaderTable), service(kj::mv(service)),
-      settings(settings), onDrain(paf.promise.fork()), drainFulfiller(kj::mv(paf.fulfiller)),
+      settings(settings), errorHandler(errorHandler.orDefault(*this)),
```
Interesting... I hadn't seen this pattern of defaulting to a privately inherited default implementation before.  I guess it's used here because it simplifies the ownership in the default case?

harrishancock replied:
Yes, exactly. I wouldn't espouse this as a pattern, but I'm lazy and it seemed to make sense. I should probably make `HttpServerErrorHandler` pure abstract anyway, though...

kentonv commented on c++/src/kj/compat/http.h:
```diff
@@ -769,21 +769,47 @@ struct HttpServerSettings {
+  // Unlike `HttpService::request()`, when calling `response.send()` in this context, a
+  // "Connection: close" header will be added, and the connection will be closed.
+  //
+  // TODO(cleanup): It might make more sense to pass in the original exception object. But this
```
I think we should pass in the exception object, so that the handler can inspect it and decide how to log it. E.g. maybe it wants to treat FAILED, OVERLOADED, and DISCONNECTED differently. Maybe the interface should have multiple methods:

    kj::Promise<void> handleClientProtocolError(kj::StringPtr message, ...);
    kj::Promise<void> handleApplicationError(kj::Exception e, ...);

(Are all the client protocol errors 400's, or are various status codes used?)

kentonv replied:
Oh yeah, and if people want to censor stack traces then they definitely want to receive the `Exception` pre-stringification...

harrishancock replied:
I think there's just the one protocol error (bad headers), and it's a 400. In that light, the two function approach sounds great.

kentonv commented on c++/src/kj/compat/http.h:
```diff
@@ -769,21 +769,47 @@ struct HttpServerSettings {
   HttpServer(kj::Timer& timer, HttpHeaderTable& requestHeaderTable,
-             HttpServiceFactory serviceFactory, Settings settings = Settings());
+             HttpServiceFactory serviceFactory, Settings settings = Settings(),
+             kj::Maybe<HttpServerErrorHandler&> errorHandler = nullptr);
```
Can `errorHandler` be a member of the `settings` struct?

harrishancock replied:
Yes, but it will be source-incompatible because the settings objects can no longer be literals. I'm fine with that, but didn't play with it enough to know what all needs to be changed. I'll find out.

kentonv replied:
Not sure what you mean. What kind of existing code would break, exactly?

harrishancock replied:
`constexpr HttpServerSettings LOCAL_CLIENT_SETTINGS = { ... };`

Probably not a big deal -- the place we'd need to modify in Workers we'd need to modify to add the new setting anyway, I just didn't spend any time seeing where else that pattern is used. I agree it'd be preferable as a setting field.

harrishancock replied:
(I'm on the bus right now, so not really explaining well.)

kentonv replied:
Is this just because `Maybe<T&>`'s default constructor is not `constexpr`? Could we make it `constexpr`?

harrishancock commented:
Okay, I think I addressed all feedback in this rewrite.

I ended up adding a third error handler function, `handleNoResponse()`, though I'm not sure it deserves to be separate from `handleApplicationError()`. I wrote it because the error message that gets generated sounds awkward if we re-use `handleApplicationError()` (something like "service threw exception: service didn't respond").

Also, should we keep the current KJ_LOG behavior in the default error handler?

harrishancock commented:
Cygwin failure looks transient.

kentonv commented on c++/src/kj/compat/http.h:
```diff
@@ -767,6 +769,45 @@ struct HttpServerSettings {
+class HttpServerErrorHandler {
+public:
+  virtual kj::Promise<void> handleClientProtocolError(kj::StringPtr message,
+      kj::HttpHeaderTable& headerTable, kj::HttpService::Response& response);
```
I think the header table shouldn't be passed as a method parameter. The normal idiom is to pass it into the constructor. E.g. HttpService::request() doesn't pass the header table; the HttpService is expected to have received it at constructor time.

harrishancock replied:
I agree, but then we'll have to make these all pure virtual, with no default implementation. If we want to allow users to call a default implementation, we could provide free functions which take a header table as a parameter, perhaps?

kentonv replied:
Could the default implementations use a default empty header table? I don't think there's actually any requirement to use the same header table as other responses. In fact, the HttpHeaderTable passed to HttpService is only meant to be used when parsing requests; I think the application can technically use whatever table it wants for responses.

harrishancock replied:
Great idea, that solves it.

kentonv commented on c++/src/kj/compat/http.c++:
```diff
@@ -4968,21 +4947,22 @@ private:
     currentMethod = nullptr;
```
It appears that every branch below now either re-assigns currentMethod back to its original value, except the one branch that sets it to GET. Maybe remove this line and all the `currentMethod = method` lines below.

harrishancock replied:
Maybe we should `currentMethod = HttpMethod::GET` here, actually? That would allow us to return an error response correctly if someone tried to open a WebSocket connection with a HEAD request (admittedly unlikely).

kentonv replied:
The very first check below this line verifies that the method is GET, so I think that's redundant. (In fact I guess this means that *all* of the later assignments to `currentMethod` can be deleted if we remove this line that sets it null.)

kentonv replied:
(And note that it would be incorrect to respond to a HEAD request with a bodied response even if the HEAD request is an invalid WebSocket handshake. So no, we should not change HEAD to GET.)

harrishancock replied:
Got it!

kentonv commented on c++/src/kj/compat/http.c++:
```diff
@@ -5147,4 +5150,76 @@ void HttpServer::taskFailed(kj::Exception&& exception) {
+        .attach(kj::mv(errorMessage), kj::mv(body));
+  }
+
+  // TODO(now): Remove this KJ_LOG? If the application wants to log this stuff, it can do it now.
```
Nah I think we should leave it. It's the right default behavior.

kentonv commented on c++/src/kj/compat/http-test.c++:
```diff
@@ -2622,6 +2622,34 @@ KJ_TEST("HttpServer cancels request when client disconnects") {
   cancelPromise.wait(waitScope);
 }
 
+KJ_TEST("HttpServer responds with 400 Bad Request") {
```
No tests that actually set a custom error handler?

harrishancock replied:
Will write one.

harrishancock commented:
Addressed everything but constexpr Maybes and lack of tests. Will pick those up in a bit.

New commits are intended as fixups, though I didn't explicitly label them such.

harrishancock commented:
Now msvc has a seemingly transient failure. Will poke it later.

harrishancock commented:
Squashed existing fixups, added tests, added `constexpr Maybe<T&>` constructors (`Maybe<T>` cannot be, due to its non-trivial destructor). @kentonv, PTAL.

kentonv commented:
LGTM but there are build failures.

harrishancock commented:
Build succeeded before fixup squash, got go-ahead from @kentonv via chat to merge.
